// Cloud Scheduler Service - Fleet-wide job scheduling
// Provides gRPC API for creating and managing scheduled jobs on vehicles
syntax = "proto3";

package ifex.cloud.scheduler;

import "scheduler-sync-envelope.proto";

// === Job Status (extended for cloud) ===

enum CloudJobStatus {
    CLOUD_JOB_UNKNOWN = 0;
    CLOUD_JOB_PENDING = 1;          // Job created, waiting to run
    CLOUD_JOB_SCHEDULED = 2;        // Job scheduled on vehicle
    CLOUD_JOB_RUNNING = 3;          // Currently executing
    CLOUD_JOB_COMPLETED = 4;        // Finished successfully
    CLOUD_JOB_FAILED = 5;           // Execution failed
    CLOUD_JOB_CANCELLED = 6;        // Job cancelled
    CLOUD_JOB_PAUSED = 7;           // Temporarily paused
}

// === Job Info ===

message JobInfo {
    string vehicle_id = 1;
    string fleet_id = 2;
    string region = 3;
    string job_id = 4;
    string title = 5;
    string service = 6;
    string method = 7;
    string parameters_json = 8;
    CloudJobStatus status = 9;
    string scheduled_time = 10;        // ISO8601 format
    string recurrence_rule = 11;       // iCal RRULE format
    string next_run_time = 12;
    uint64 created_at_ms = 13;
    uint64 updated_at_ms = 14;
    string created_by = 15;            // Who created this job
}

// === Execution Info ===

message ExecutionInfo {
    string execution_id = 1;
    CloudJobStatus status = 2;
    uint64 executed_at_ms = 3;
    uint32 duration_ms = 4;
    string result_json = 5;
    string error_message = 6;
    string next_run_time = 7;
}

// === Create Job ===

message CreateJobRequest {
    string vehicle_id = 1;
    string title = 2;                  // Human-readable title
    string service = 3;                // Target service name
    string method = 4;                 // Method to invoke
    string parameters_json = 5;        // JSON-encoded parameters
    string scheduled_time = 6;         // When to first run (ISO8601)
    string recurrence_rule = 7;        // Optional: cron/iCal RRULE for recurring
    string end_time = 8;               // Optional: when to stop recurring
    string created_by = 9;             // Who is creating this job
}

message CreateJobResponse {
    bool success = 1;
    string job_id = 2;                 // Cloud-generated job ID
    string error_message = 3;
}

// === Update Job ===

message UpdateJobRequest {
    string vehicle_id = 1;
    string job_id = 2;
    string title = 3;                  // Empty = no change
    string scheduled_time = 4;         // Empty = no change
    string recurrence_rule = 5;        // Empty = no change
    string parameters_json = 6;        // Empty = no change
    string end_time = 7;               // Empty = no change
}

message UpdateJobResponse {
    bool success = 1;
    string error_message = 2;
}

// === Delete Job ===

message DeleteJobRequest {
    string vehicle_id = 1;
    string job_id = 2;
}

message DeleteJobResponse {
    bool success = 1;
    string error_message = 2;
}

// === Pause/Resume Job ===

message PauseJobRequest {
    string vehicle_id = 1;
    string job_id = 2;
}

message PauseJobResponse {
    bool success = 1;
    string error_message = 2;
}

message ResumeJobRequest {
    string vehicle_id = 1;
    string job_id = 2;
}

message ResumeJobResponse {
    bool success = 1;
    string error_message = 2;
}

// === Trigger Job (immediate execution) ===

message TriggerJobRequest {
    string vehicle_id = 1;
    string job_id = 2;
}

message TriggerJobResponse {
    bool success = 1;
    string error_message = 2;
}

// === Get Job ===

message GetJobRequest {
    string vehicle_id = 1;
    string job_id = 2;
}

message GetJobResponse {
    bool found = 1;
    JobInfo job = 2;
}

// === List Jobs ===

message ListJobsRequest {
    string vehicle_id_filter = 1;      // Optional
    string fleet_id_filter = 2;        // Optional
    string region_filter = 3;          // Optional
    string service_filter = 4;         // Optional
    CloudJobStatus status_filter = 5;  // Optional (0 = all)
    string created_by_filter = 6;      // Optional
    bool recurring_only = 7;           // Only recurring jobs
    int32 page_size = 10;
    string page_token = 11;
}

message ListJobsResponse {
    repeated JobInfo jobs = 1;
    string next_page_token = 2;
    int32 total_count = 3;
}

// === Get Job Executions ===

message GetJobExecutionsRequest {
    string vehicle_id = 1;
    string job_id = 2;
    int32 limit = 3;                   // Default: 100
    uint64 since_ms = 4;               // Optional: only executions after this time
}

message GetJobExecutionsResponse {
    string vehicle_id = 1;
    string job_id = 2;
    repeated ExecutionInfo executions = 3;
    int32 total_count = 4;
}

// === Create Fleet Job (same job on multiple vehicles) ===

message CreateFleetJobRequest {
    repeated string vehicle_ids = 1;   // Explicit list of vehicles
    string fleet_id = 2;               // OR all vehicles in fleet
    string region = 3;                 // OR all vehicles in region
    string service_filter = 4;         // OR only vehicles with this service
    string title = 5;
    string service = 6;
    string method = 7;
    string parameters_json = 8;
    string scheduled_time = 9;
    string recurrence_rule = 10;
    string end_time = 11;
    string created_by = 12;
}

message FleetJobResult {
    string vehicle_id = 1;
    bool success = 2;
    string job_id = 3;
    string error_message = 4;
}

message CreateFleetJobResponse {
    int32 total_vehicles = 1;
    int32 successful = 2;
    int32 failed = 3;
    repeated FleetJobResult results = 4;
    string fleet_job_id = 5;           // ID to track the fleet operation
}

// === Delete Fleet Job ===

message DeleteFleetJobRequest {
    repeated string vehicle_ids = 1;
    repeated string job_ids = 2;       // Delete these jobs from all specified vehicles
}

message DeleteFleetJobResponse {
    int32 total_deletions = 1;
    int32 successful = 2;
    int32 failed = 3;
}

// === Get Fleet Job Stats ===

message GetFleetJobStatsRequest {
    string fleet_id_filter = 1;
    string region_filter = 2;
}

message JobStats {
    string service = 1;
    string method = 2;
    int32 total_jobs = 3;
    int32 pending = 4;
    int32 running = 5;
    int32 completed = 6;
    int32 failed = 7;
    int32 recurring = 8;
}

message GetFleetJobStatsResponse {
    int32 total_jobs = 1;
    int32 total_vehicles_with_jobs = 2;
    repeated JobStats by_service_method = 3;
    map<string, int32> by_status = 4;
    map<string, int32> by_region = 5;
}

// === gRPC Service Definition ===

service CloudSchedulerService {
    // Single vehicle operations
    rpc CreateJob(CreateJobRequest) returns (CreateJobResponse);
    rpc UpdateJob(UpdateJobRequest) returns (UpdateJobResponse);
    rpc DeleteJob(DeleteJobRequest) returns (DeleteJobResponse);
    rpc PauseJob(PauseJobRequest) returns (PauseJobResponse);
    rpc ResumeJob(ResumeJobRequest) returns (ResumeJobResponse);
    rpc TriggerJob(TriggerJobRequest) returns (TriggerJobResponse);

    // Query operations (read from PostgreSQL)
    rpc GetJob(GetJobRequest) returns (GetJobResponse);
    rpc ListJobs(ListJobsRequest) returns (ListJobsResponse);
    rpc GetJobExecutions(GetJobExecutionsRequest) returns (GetJobExecutionsResponse);

    // Fleet-wide operations
    rpc CreateFleetJob(CreateFleetJobRequest) returns (CreateFleetJobResponse);
    rpc DeleteFleetJob(DeleteFleetJobRequest) returns (DeleteFleetJobResponse);
    rpc GetFleetJobStats(GetFleetJobStatsRequest) returns (GetFleetJobStatsResponse);
}
